main.go

```diff
@@ -0,0 +1,273 @@
+	getterCtx := getterContext{rw, req, next}
+	ctx := context.WithValue(req.Context(), getterContextCtxKey, getterCtx)
+	req = req.WithContext(ctx)
+	buf := s.bufPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer s.bufPool.Put(buf)
+	combo := ctx.Value(getterContextCtxKey).(getterContext)
+	plugins.DefaultSouinPluginCallback(rw, req, s.Retriever, s.RequestCoalescing, func(_ http.ResponseWriter, _ *http.Request) error {
+		recorder := httptest.NewRecorder()
+		e := combo.next.ServeHTTP(recorder, combo.req)
+		if e != nil {
+			return e
+		}
+
+		response := recorder.Result()
+		req.Response = response
+		response, e = s.Retriever.GetTransport().(*rfc.VaryTransport).UpdateCacheEventually(req)
+		if e != nil {
+			return e
+		}
+
+		_, e = io.Copy(rw, response.Body)
+
+		return e
+	})
+
+	return nil
+}
+
+func (s *SouinCaddyPlugin) configurationPropertyMapper() error {
+	if s.Configuration != nil {
+		return nil
+	}
+	defaultCache := &DefaultCache{
+		Distributed: s.Olric.URL != "" || s.Olric.Path != "" || s.Olric.Configuration != nil,
+		Headers:     s.Headers,
+		Olric:       s.Olric,
+		TTL:         s.TTL,
+	}
+	if s.Configuration == nil {
+		s.Configuration = &Configuration{
+			DefaultCache: defaultCache,
+			LogLevel:     s.LogLevel,
+		}
+	}
+	s.Configuration.DefaultCache = defaultCache
+	return nil
+}
+
+// Validate to validate configuration.
+func (s *SouinCaddyPlugin) Validate() error {
+	return nil
+}
+
+// FromApp to initialize configuration from App structure.
+func (s *SouinCaddyPlugin) FromApp(app *SouinApp) error {
+	if s.Configuration == nil {
+		s.Configuration = &Configuration{
+			URLs: make(map[string]configurationtypes.URL),
+		}
+	}
+
+	if app.DefaultCache == nil {
+		return nil
+	}
+
+	if s.Configuration.DefaultCache == nil {
+		s.Configuration.DefaultCache = &DefaultCache{
+			Headers:     app.Headers,
+			TTL:         app.TTL,
+		}
+	} else {
+		dc := s.Configuration.DefaultCache
+		appDc := app.DefaultCache
+		if dc.Headers == nil {
+			s.Configuration.DefaultCache.Headers = appDc.Headers
+		}
+		if dc.TTL == "" {
+			s.Configuration.DefaultCache.TTL = appDc.TTL
+		}
+		if dc.Olric.URL == "" || dc.Olric.Path == "" || dc.Olric.Configuration == nil {
+			s.Configuration.DefaultCache.Distributed = appDc.Distributed
+			s.Configuration.DefaultCache.Olric = appDc.Olric
+		}
+	}
+
+	return nil
+}
+
+// Provision to do the provisioning part.
+func (s *SouinCaddyPlugin) Provision(ctx caddy.Context) error {
+	s.logger = ctx.Logger(s)
+
+	if err := s.configurationPropertyMapper(); err != nil {
+		return err
+	}
+
+	app, _ := ctx.App(moduleName)
+
+	if err := s.FromApp(app.(*SouinApp)); err != nil {
+		return err
+	}
+
+	s.bufPool = sync.Pool{
+		New: func() interface{} {
+			return new(bytes.Buffer)
+		},
+	}
+	s.Retriever = plugins.DefaultSouinPluginInitializerFromConfiguration(s.Configuration)
+	s.RequestCoalescing = coalescing.Initialize()
+	return nil
+}
+
+func parseCaddyfileRecursively(h *caddyfile.Dispenser) interface{} {
```

Comment:
This doesn't look like a good approach. I'd call this a hack more than anything.

I think the configuration should be more explicitly done rather than relying on the existence of `}` values. No other Caddyfile parsing does that, because that should be abstracted away.

Reply:
This is done to load the Olric configuration from the Caddyfile without duplicating the Olric internal type and map values over this.